README

This document describes major functions of the driver (universal-driver-8b8)
deployed in Niagara Bypass NIC cards with either copper or fiber interfaces.

The Niagara Bypass NIC card can monitor its host workstation and automatically
switch Ether interface to BYPASS when the watchdog heartbeat from the host is
absent.
  
As of writing this document, the driver version is 8b8. The driver is
written under GPL license.

1. Software requirements:
	- Linux version 2.6.27 or higher
	- for compilation: installed gnu make,gcc, libc-dev,
		linux-kernel-headers,awk
	- for launch: installed udev (to manage device /dev/niagara
		mode and permission)
	- Ununtu suggested
2. Supported cards
	- cards with USB or ETH control interfaces are not supported, only
		PCI/PCIe
	- any fiber/copper Niagara NIC with Intel MAC, which supports
		"software defineable pin side interface". For example, N32264
		or N42264.

3. Quick start
3.1. obtain universal-driver-8b8, unpack, compile
	$ make
		expected no error/warning messages, driver modulo and utility
		will be built
	$ sudo make install
		niagara_util will be placed in /usr/local/bin
	$ make insmod
		modulo launched. You can check dmesg or /var/log/kern.log
3.2. Play with niagara_util.
List of detected devices:
	$ niagara_util -Sv
	0       N32264 ( 4 ports COPPER 1G IGB SDP )
Check for all options. Call without arguments will show usage.
If card or segment is not set - it means all of them:
	$ niagara_util 
	niagara_util [OPTIONS] [card[.segment]]
        	-S      Count cards
        	-i      Card/segment information
	Segment mode
        	-d<0..5>Set Default mode
                        	HeartBeatPresent NoHeartBeat
                Copper modes:
                        -d0       Active          Bypass
                        -d1       Bypass          Active
                        -d2       Active          Active
                        -d3       Active          NoLink
                        -d4       Bypass          Bypass
                        -d5       Nolink          Nolink
                Fiber modes:
                        -d0       Active          Active
                        -d1       Bypass          Bypass
                        -d2       Active          Bypass
                        -d3       Bypass          Active
                        -d4       Active          NoLink
                        -d5       NoLink          NoLink
                -c<0..5>Set Current mode
                -cd     Set Current mode to Default
	Segment watchdog
        	-t<1..ff>Set timeout (in 0.1 sec)
        	-k      Kick segment
        	-a      Add heartbeat
        	-r      Remove heartbeat
        	-h<1..ff>Heartbeat period (in 0.1 sec)
	Link Fault Detection
        	-Q      Enable LFD
        	-q      Disable LFD
	Power Off mode
		-P	Bybass
		-p	Regular
	Read/write to register/memory on local CPU (in cpld_debug mode only!)
	-g <a>	a is 1-byte long hexadecimal number -
		address on local CPU to read
	-s <ad>	ad is 2-byte long hexadecimal number -
		first byte is address, second - data to write
	Common options
        	-v      verbose
	EVERYTHING IS HEXADECIMAL EVERYWHERE!
3.3. Get information about the first segment of the first card.
	$ niagara_util -i 0.0
	0.0     TIMEOUT:10
	0.0     RELAY:01        BYPASS
	0.0     CURRENT MODE:70
	0.0     DEFAULT MODE:00
	0.0     HEARTBEAT:00
     meaning of fields are described later, now You can see, that card is in
     	BYPASS mode. 
     Also You can get information about all segments of the first card:
	$ niagara_util -i 0
     Or about all cards:
        $ niagara_util -i
3.4. Switching on monitoring. With this call we are asking kernel periodically
"kicks" segment with keep-alive signals.
	$ niagara_util -a 0.0
While receiving this kicks, named heartbeat, card thinks, that everything is
OK and switched to ACTIVE mode:
	$ niagara_util -i 0.0
	0.0     TIMEOUT:10
	0.0     RELAY:02        ACTIVE
	0.0     CURRENT MODE:70
	0.0     DEFAULT MODE:00
	0.0     HEARTBEAT:01    PERIOD:0F
We can combine switches in one line, they have internal priority, -i is
executed last:
	$ niagara_util -i 0.0 -a
3.5. Playing with /proc filesystem. All parameters are represented there,
most of them can be changed:
	$ ls -R /proc/niagara/
	/proc/niagara/:
	0
	/proc/niagara/0:
	0  1  cpld_id  firmware_rev  flags  name  num_ports  oem_id
		product_id  product_rev  secondary_firmware_rev
	/proc/niagara/0/0:
	current_mode  default_mode  heartbeat  heartbeat_period
		relay_status  timeout_scaler
	/proc/niagara/0/1:
	current_mode  default_mode  heartbeat  heartbeat_period  relay_status
		timeout_scaler
Let's check heartbeat status of segment 0.0:	
	$ cat /proc/niagara/0/0/heartbeat
	01
Let's stop it:
	$ echo 0 > /proc/niagara/0/0/heartbeat
	$ cat /proc/niagara/0/0/heartbeat
	00
Let's see segment status with utility. It's switched to BYPASS - our watchdog
works:
	$$ niagara_util -i 0.0
	0.0     TIMEOUT:10
	0.0     RELAY:01        BYPASS
	0.0     CURRENT MODE:70
	0.0     DEFAULT MODE:00
	0.0     HEARTBEAT:00
The same information can be seen with procfs:
	$ cat /proc/niagara/0/0/relay_status 
	01

4. Base terms
We have only three types of objects: cards,segments and ports.
Card is physical PCB, inserted into PCI. It consists of segemnts, one or more.
Each segment has exactly two ports.
Each object has attributes, which can be readonly or write-capable.
Segments are totally logically  independend, despite placed in one PCB.
Internally all segments in one PCB are served by one CPU, that's why driver
(internally) uses locking technique to avoid concurrent access to CPU.

5. Attribute meaning
5.1 Each card has 11 attributes:
	cpld_id (RO)
		two hexa-digit displays board revision and CPLD program
		version. You can ignore it ;)
	firmware_rev (RO)
		self-identification register of card's CPU
	flags (RO)
		this sophistical bits explain hardware features of this card.
		bit descriptions can be found in include/niagara_flags.h
	name (RO)
		commercial name of this card
	num_ports (RO)
		number of ports (not segments!)
	oem_id (RO)
		constant value, reported by CPU
		Should be 26 for Niagara products.
	product_id (RO)
		Manufacturer's number of this device.
		List of possible ID's can be found outside this document
	product_rev (RO)
		product variant
	secondary_firmware_rev (RO)
		reserved identification byte, 01 for all current products
	avr_sig (RO)
		AVR CPU signature. Should be 1E9307 for ATMega8
5.2. Each segment has 8 attributes:
	current_mode (rw)
		this byte describes what segment shell to do in different
		situations.
		There are preset six different scenarios wor copper card,
		please see list in  niagara_util usage.
		Special value 70 (default) means redirection, mode will be
		taken from default byte.
		Note, that current_mode will NOT kept after device reboot.
	default_mode (rw)
		value can be 0..5, the meaning is same as above.
		This byte is kept while reboot.
	heartbeat (rw)
		can be 00 - no heartbeat or 01 - heartbeat active.
	heartbeat_period (rw)
		in 0.1 secs
	relay status (ro)
		01 BYPASS
		02 ACTIVE
		03 NOLINK
	timeout_scaler (rw)
		in 0.1 secs, this value is an instruction for internal card's
		watchdog.
		It's recommended have timeout_scaler about 1.5 times greater
		then heartbit_period
	LFD (Link Fault Detection) (rw)
		LFD is when one port is down, its companion port will be
		brought down.  However a copper NIC presents its interface
		status slightly different than a fiber NIC when the LFD is
		turned on.  This difference can be observed in port LEDS or
		using niagara_util.  A copper NIC would turn off both port
		LEDs and display both port status as “00”.  The fiber NIC
		would only turn off the LED on the defunct port, display its
		port status as “00” and leave its companion port display
		alone. Use “ethtool ethXX” at the CLI to verify interface
		status is recommended in case like this.
       Fail/Power off Mode (rw)
		Power off mode dictates if the NIC will fail open or fail
		close. Fail open (bypass) will allow traffic to flow when
		power is lost or there is a software failure. Fail close
		(regular) will not allow traffic to flow when power is lost
		or there is a software failure. This applies only to copper
		NICs. This mode does not apply to Fiber NICs.

5.3 Port attributes is not implemented in this version
6. Software parts
   Internally driver consists of:
	- cpld layer
		low-level functions, which deals with Intel MAC chip.
		The main tasks - just send or recieve bits to card's CPLD.
		In 0.1 version of driver only one type of access implemented,
		SDP.
	- cpu layer
		this layer allows send/recieve bytes to card's CPU.
		locking technique is used here.
	- pci enumeration
		this part of software knows "our" pci devices and performs
		their initialisation.
		list of devices can be found in human-readable file
		"supported_cards.txt".
		Implementation in version 0.1 are based on pci_scan approch.
		Full-featured "Linux Device Model" is planned in Version 0.2
	- kernel_api
		This is a set of procedures for manipulation with attributes.
		It can be called from any part of kernel, even from
		third-party driver
	- procfs wrapper
		when someone read or write files in /proc/niagara, this
		wrapper translate it to kernel_api calls
	- sysfs wrapper
		the same with /sys/devices/virtual/niagara, in version 0.1.
		implemented only for card attributes
	- user_api
		user-space library for manipulation with attributes
	- niagara_test
		simple utility, which just calls user_api library and shows
		results
7. Dixi.


Changelog
Version 0.8b8 20162511
	- resolved sporadical module lock on exit issue
	- resolved LFD thread lock on port state update issue

Version 0.8b7 20160408
	- resolved issues with port status mismatch between segment interfaces for N32285 and N32285
	- resolved LFD functionality issues for N32284 and N32285

Version 0.8b6 20160607
	- bugfixes for N32285: LFD issue (#2690) and link status mismatch (#2691)

Version 0.8b5 20160604
	- resolved issue with swapped segment registers for N42264 and N32264
	- resolved few minor issues with sysfs attributes permissions and README

Version 0.8 b4 20150712
	- resolved module compillation issue for Linux kernels 3.18+

Version 0.8 b3 20141812
	- fixed LFD functionality for some types of NICS with swapped SPD/ESPD 
	  control registers (N32265 for example)

Version 0.8 b2 20141908
	- fixed set MTU size isue for i350 based NICs
	- fixed very rare system hangs caused by universal driver
	- fixed incorrect port status information in some modes for fiber NICs

Version 0.8 b1 20142807
	- fixed deadlock in 'kick' functionality for MDIO based NICs
	- fixed LFD functionality for FIBER NICs in default mode
	- fixed 'mismatched interfaces' issue for some NICs
	- added support for i350 based NICs (N52284/N52285)

Version 0.7 b4 20140706
	- fixed LFD functionality for fiber NICs
	- added ability to enable/disable CPLD debug mode
		- access to CPLD registers via sysfs
		- communicate with local CPU via sysfs
	- added ability to comunicate to local CPU using niagara_util

Version 0.7 b3 20140606
	- added ability to set power off mode (bypass or regular)

Version 0.7 b2 20143001
	- support for new procfs kernel APIs (kernel 3.10 and upper)
	- workaround for ethernet controller resets caused by ixgbe driver (N32710-TX)

Version 0.7 b1 20131105
	- port status obtained from net_device abstraction
	- LFD for N32710-TX implemented (no access to PHY)

Version 0.6 b2 20131012
	- kernel API implemented
	- example API files provided
	
Version 0.6 b1 20130607
	- fix kick() if card is busy, not poll but retry later

Version 0.5 b1 20130410
	- verbose output while insmod
	- prohibition of unloading Intel driver
	- card binary attributes cpld_mem and cpu_mem

Version 0.3.b1 20130303
	- FLASH cards fully supported (N2282, N2261E)
	- kick(): unneeded locking removed

Version 0.2.b1 20130103
	- sysfs implemented in full volume
	- API changed (from structures to attributes)
	- sysfs used as main wrapper
	- ioctl access removed

Version 0.1.b2 20121230
	- README added
	- versioning support, check the same driver/utility version
	- /proc/niagara/version added
	- utility: verbose output for CURRENT_MODE 0x70
	- uncrustify sources
	- api: TIMEOUT 00 prohibited
	- /proc/niagara/supported_cards added
	- niagara_util -l 


user_api/util.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>

#include <niagara_api.h>
#include <niagara_flags.h>


#define error(args ...) do { fprintf(stderr, "ERROR "); fprintf(stderr, args); fprintf(stderr, errno ? ":%m\n" : "\n"); exit(EXIT_FAILURE); } while (0)


static inline const char *relay_status2str(unsigned char x)
{
	switch (x) {
	case 1: return "BYPASS";
	case 2: return "ACTIVE";
	case 3: return "NOLINK";
	default: return "XXX";
	}
}

static void usage(void)
{
	printf("Niagara BYPASS NIC support utility Version %d.%db%d\n"
	       "niagara_util [OPTIONS] [card[.segment]]\n"
	       "\t-l\tList of supported cards\n"
	       "\t-S\tCount installedcards\n"
	       "\t-i\tCard/segment information\n"
	       "Segment mode\n"
	       "\t-d<0..5>Set Default mode\n"
	       "\t\t\tHeartBeatPresent NoHeartBeat\n"
	       "\tCopper modes:\n"
	       "\t\t-d0\t  Active\t  Bypass\n"
	       "\t\t-d1\t  Bypass\t  Active\n"
	       "\t\t-d2\t  Active\t  Active\n"
	       "\t\t-d3\t  Active\t  NoLink\n"
	       "\t\t-d4\t  Bypass\t  Bypass\n"
	       "\t\t-d5\t  Nolink\t  NoLink\n"
	       "\tFiber modes:\n"
	       "\t\t-d0\t  Active\t  Active\n"
	       "\t\t-d1\t  Bypass\t  Bypass\n"
	       "\t\t-d2\t  Active\t  Bypass\n"
	       "\t\t-d3\t  Bypass\t  Active\n"
	       "\t\t-d4\t  Active\t  NoLink\n"
	       "\t\t-d5\t  Nolink\t  Nolink\n"
	       "\t-c<0..5>Set Current mode\n"
	       "\t-cd\tSet Current mode to Default\n"
	       "Segment watchdog\n"
	       "\t-t<1..ff>Set timeout (in 0.1 sec)\n"
	       "\t-k\tKick segment\n"
	       "\t-a\tAdd heartbeat\n"
	       "\t-r\tRemove heartbeat\n"
	       "\t-h<1..ff>Heartbeat period (in 0.1 sec)\n"
	       "Link Fault Detection\n"
	       "\t-Q\tEnable LFD\n"
	       "\t-q\tDisable LFD\n"
	       "Power Off mode\n"
	       "\t-P\tBypass\n"
	       "\t-p\tRegular\n"
	       "Read/write to register/memory on local CPU (in CPLD debug mode only!)\n"
	       "\t-g <a>\ta is 1-byte long hexadecimal number -\n\t\taddress on local CPU to read\n"
	       "\t-s <ad>\tad is 2-byte long hexadecimal number -\n\t\tfirst byte is address, second - data to write\n"
	       "Common options\n"
	       "\t-v\tverbose\n"
	       "EVERYTHING IS HEXADECIMAL EVERYWHERE!\n",
	       NIAGARA_VERSION_MAJOR,
	       NIAGARA_VERSION_MINOR,
	       NIAGARA_VERSION_BUILD
	       );
	exit(EXIT_FAILURE);
}

static int verbose = 1;
static unsigned cmd = 0;
#define CMD_CARD_LIST   (1 << 0)
#define CMD_CARD_COUNT  (1 << 1)
#define CMD_INFO        (1 << 2)
#define CMD_TIMEOUT     (1 << 3)
#define CMD_KICK        (1 << 4)
#define CMD_ADD_HEART   (1 << 5)
#define CMD_DEL_HEART   (1 << 6)
#define CMD_PERIOD      (1 << 7)
#define CMD_C           (1 << 8)
#define CMD_D           (1 << 9)
#define CMD_ADD_LFD     (1 << 10)
#define CMD_DEL_LFD     (1 << 11)
#define CMD_PWR_OFF_BP	(1 << 12)
#define CMD_PWR_OFF_RG	(1 << 13)
#define CMD_CPU_READ	(1 << 14)
#define CMD_CPU_WRITE	(1 << 15)

static unsigned maxcard;
static int card, segment;

static int getarg(int *arg)
{
	char *eptr;

	*arg = strtoul(optarg, &eptr, 16);
	if (eptr == optarg) return -1;
	if (*eptr != 0) return -1;
	return 0;
}

int main(int argc, char **argv)
{
	int opt, i, j;
	int timeout_arg, c_arg, d_arg, h_arg, cpu_read_arg, cpu_write_arg;

	while ((opt = getopt(argc, argv, "Slit:d:c:karh:QqPpg:s:v")) != -1)
		switch (opt) {
		default: usage();
		case 'l': cmd |= CMD_CARD_LIST; continue;
		case 'S': cmd |= CMD_CARD_COUNT; continue;
		case 'i': cmd |= CMD_INFO; continue;
		case 't': if (getarg(&timeout_arg)) usage();
			if (timeout_arg > 255) usage();
			cmd |= CMD_TIMEOUT; continue;
		case 'd': if (getarg(&d_arg)) usage();
			cmd |= CMD_D; continue;
		case 'c': if (strcmp(optarg, "d") == 0) c_arg = 0x70;
			else if (getarg(&c_arg)) usage();
			cmd |= CMD_C; continue;
		case 'k': cmd |= CMD_KICK; continue;
		case 'a': cmd |= CMD_ADD_HEART; continue;
		case 'r': cmd |= CMD_DEL_HEART; continue;
		case 'h': if (getarg(&h_arg)) usage();
			cmd |= CMD_PERIOD; continue;
		case 'Q': cmd |= CMD_ADD_LFD; continue;
		case 'q': cmd |= CMD_DEL_LFD; continue;
		case 'P': cmd |= CMD_PWR_OFF_BP; continue;
		case 'p': cmd |= CMD_PWR_OFF_RG; continue;
		case 'g': if (getarg(&cpu_read_arg)) usage();
			if (cpu_read_arg > 255) usage();
			cmd |= CMD_CPU_READ; continue;
		case 's': if (getarg(&cpu_write_arg)) usage();
			if (cpu_write_arg > 0xffff) usage();
			cmd |= CMD_CPU_WRITE; continue;
		case 'v': verbose++; continue;
		}
	if (optind != argc) {
		if (cmd & (CMD_CARD_COUNT | CMD_CARD_LIST)) usage();
		if (sscanf(argv[optind], "%x.%x", &card, &segment) == 2) {
			optind++;
		} else if (sscanf(argv[optind], "%x", &card) == 1) {
			segment = -1;
			optind++;
		} else { usage(); }
	} else {
		card = -1;
		segment = -1;
	}
	if (optind != argc) usage();
	if (cmd == 0) usage();
	{ unsigned version;
	  NiagaraGetAttribute(0, 0, "version", &version);
	  if (NIAGARA_VERSION != version) error("Utility version and driver version mismatch"); }
	if (cmd & CMD_CARD_LIST)
		system("cat /sys/devices/virtual/niagara/supported_cards");
	NiagaraGetAttribute(0, 0, "maxcard", &maxcard);
	if (cmd & CMD_CARD_COUNT) {
		if (verbose <= 1) {
			printf("%X\n", maxcard);
			return 0;
		}
		for (i = 0; i < maxcard; i++) {
			unsigned num_ports, flags;
			char cardname[64];
			NiagaraGetAttribute(i, 0, "num_ports", &num_ports);
			NiagaraGetAttribute(i, 0, "flags", &flags);
			NiagaraGetCharAttribute(i, 0, "name", cardname, sizeof(cardname));
			printf("%X\t%s ( %d ports %s )\n", i, cardname, num_ports, flag2str(flags));
		}
	}
	if ((card != -1) && (card >= maxcard))
		error("Invalid card number , %x max", maxcard - 1);
	for (i = 0; i < maxcard; i++)
		if ((card == -1) || (card == i)) {
			unsigned num_ports;
			NiagaraGetAttribute(i, 0, "num_ports", &num_ports);
			if ((card != -1) && (segment != -1) && (segment >= num_ports / 2))
				error("Invalid segment number , %x.%x max", card, num_ports / 2 - 1);
			if (verbose > 0) {
				if ((cmd & CMD_INFO) && (segment == -1)) {
					unsigned oem_id, cpld_id, product_id, product_rev, firmware_rev, secondary_firmware_rev;
					char cardname[64];
					NiagaraGetAttribute(i, 0, "oem_id", &oem_id);
					NiagaraGetAttribute(i, 0, "cpld_id", &cpld_id);
					NiagaraGetAttribute(i, 0, "product_id", &product_id);
					NiagaraGetAttribute(i, 0, "product_rev", &product_rev);
					NiagaraGetAttribute(i, 0, "firmware_rev", &firmware_rev);
					NiagaraGetAttribute(i, 0, "secondary_firmware_rev", &secondary_firmware_rev);
					NiagaraGetCharAttribute(i, 0, "name", cardname, sizeof(cardname));
					printf("Card #%X %s\n", i, cardname);
					printf("%X\tOEM/CPLD:%02X.%02X\n", i, oem_id, cpld_id);
					printf("%X\tPRODUCT :%02X.%02X\n", i, product_id, product_rev);
					printf("%X\tFIRMWARE:%02X:%02X\n", i, firmware_rev, secondary_firmware_rev);
				}
			}
			if (cmd & CMD_CPU_READ) {
				unsigned cpu_read_data;
				NiagaraSetAttribute(i, j, "cpld/IOREG", cpu_read_arg << 8);
				NiagaraGetAttribute(i, j, "cpld/IOREG", &cpu_read_data);
				printf("card:%d:\taddr:0x%02X\tdata:0x%02X\n", i, cpu_read_arg, 0xff & cpu_read_data);
			}
			if (cmd & CMD_CPU_WRITE)
				NiagaraSetAttribute(i, 0, "cpld/IOREG", 0x1000000 | cpu_write_arg);
			for (j = 0; j < num_ports / 2; j++)
				if ((segment == -1) || (segment == j)) {
					if (cmd & CMD_TIMEOUT)
						NiagaraSetAttribute(i, j, "timeout_scaler", timeout_arg);
					if (cmd & CMD_C)
						NiagaraSetAttribute(i, j, "current_mode", c_arg);
					if (cmd & CMD_D)
						NiagaraSetAttribute(i, j, "default_mode", d_arg);
					if (cmd & CMD_KICK)
						NiagaraSetAttribute(i, j, "kick", 1);
					if (cmd & CMD_PERIOD)
						NiagaraSetAttribute(i, j, "heartbeat_period", h_arg);
					if (cmd & CMD_ADD_HEART)
						NiagaraSetAttribute(i, j, "heartbeat", 1);
					if (cmd & CMD_DEL_HEART)
						NiagaraSetAttribute(i, j, "heartbeat", 0);
					if (cmd & CMD_ADD_LFD)
						NiagaraSetAttribute(i, j, "lfd", 1);
					if (cmd & CMD_DEL_LFD)
						NiagaraSetAttribute(i, j, "lfd", 0);
					if (cmd & CMD_PWR_OFF_BP)
						NiagaraSetAttribute(i, j, "power_off_mode", 1);
					if (cmd & CMD_PWR_OFF_RG)
						NiagaraSetAttribute(i, j, "power_off_mode", 0);
					if (cmd & CMD_INFO) {
						unsigned timeout_scaler;
						unsigned relay_status;
						unsigned current_mode;
						unsigned default_mode;
						unsigned lfd;
						unsigned port0;
						unsigned port1;
						unsigned heartbeat;
						unsigned heartbeat_period;
						unsigned power_off_mode;
						char port0_name[32] = {0};
						char port1_name[32] = {0};
						NiagaraGetAttribute(i, j, "timeout_scaler", &timeout_scaler);
						NiagaraGetAttribute(i, j, "relay_status", &relay_status);
						NiagaraGetAttribute(i, j, "current_mode", &current_mode);
						NiagaraGetAttribute(i, j, "default_mode", &default_mode);
						NiagaraGetAttribute(i, j, "lfd", &lfd);
						NiagaraGetAttribute(i, j, "port0", &port0);
						NiagaraGetAttribute(i, j, "port1", &port1);
#if !(USE_PROCFS == 1)
						NiagaraGetCharAttribute(i, j, "port0_name", port0_name, sizeof(port0_name));
						NiagaraGetCharAttribute(i, j, "port1_name", port1_name, sizeof(port1_name));
#else
						snprintf(port0_name, sizeof(port0_name), "0");
						snprintf(port1_name, sizeof(port1_name), "1");
#endif
						NiagaraGetAttribute(i, j, "heartbeat", &heartbeat);
						NiagaraGetAttribute(i, j, "heartbeat_period", &heartbeat_period);
						NiagaraGetAttribute(i, j, "power_off_mode", &power_off_mode);
						printf("%X.%X\tTIMEOUT:%02X\n", i, j, timeout_scaler);
						printf("%X.%X\tRELAY:%02X", i, j, relay_status);
						if (verbose)
							printf("\t%s\n", relay_status2str(relay_status));
						else
							printf("\n");
						printf("%X.%X\tCURRENT MODE:%02X", i, j, current_mode);
						if (current_mode == 0x70)
							printf("\tDEFAULT\n");
						else
							printf("\n");
						printf("%X.%X\tDEFAULT MODE:%02X\n", i, j, default_mode);
						printf("%X.%X\tLFD:%02X PORT0(%s):%02X PORT1(%s):%02X\n", i, j, lfd, port0_name, port0, port1_name, port1);
						printf("%X.%X\tHEARTBEAT:%02X", i, j, heartbeat);
						if (heartbeat)
							printf("\tPERIOD:%02X\n", heartbeat_period);
						else
							printf("\n");
						printf("%X.%X\tPOWER OFF MODE: %s\n", i, j,
							power_off_mode?"BYPASS":"REGULAR");
					}
				}
		}
	return 0;
}
